### DISCUSSION TITLE
Applied current [A] - function of time

### DISCUSSION THREAD

**Discussion by veera-visw** [pybamm-team/PyBaMM#3863] [Q&A] [answered]
I am working on an extension of the half cell example case, including some gas generation reactions. I now have a stable code and I wish to include some battery cycling in this simulation.
Right now, the way the applied curernt is coded in is as follows:

param = pybamm.ParameterValues(
    {
.....
        "Applied current [A]": -1,
......
    }
)
//Solution
solver = pybamm.CasadiSolver("fast",root_tol=3e-3)
t_eval = np.linspace(0, 3600, 1000)
solution = solver.solve(model, t_eval)

In this environment, is it possible to replace the constant value of current with a variable current as a function of pybamm.t? 
I tried the following:
i=np.copysign(1, np.sin(2*np.pi*pybamm.t*f))

and use "i" as an input. But I get the following error:
AttributeError: module 'pybamm' has no attribute 'copysign'

Any suggestions? Need not be a direct one, I need a simple way to finish charging cycle and use the solution as an initialization for discharge. At the moment, I am result oriented so not looking for pretty solutions.

**Comment by valentinsulzer** [comment, marked as answer]
Try something like `i = (pybamm.t < 3600) * (-5) + (pybamm.t >= 3600) * (pybamm.t < 7200) * 5` 
That would be 1h 1C charge followed by 1h 1C charge for a 5 Ah cell.
Also, we are aiming to have [experiments](https://docs.pybamm.org/en/stable/source/examples/notebooks/getting_started/tutorial-5-run-experiments.html) work with any model in the next release

**Comment by veera-visw** [comment]
@tinosulzer  That was helpful. Thanks and good to know that the experiment class will be extended to any models. For a beginner like me, it is much easier to play around with the BaseModel(), so I am happy about this.

**Comment by veera-visw** [comment]
@valentinsulzer  I have a follow up on the same topic - just avoiding the opening of a new discussion. 

As an extension, I wish to do the applied current based on a voltage cut off: Basically, charge until a certain voltage is reached and then discharge. 

Following is my attempt:

BV=pybamm.boundary_value(phi_p, "right")-pybamm.boundary_value(phi_n, "left")
voltage_check= BV-high_cutoff
i_applied = (voltage_check<=0)* (-curr*1) + (voltage_check>0)*(curr*0.25)

I am getting the following error and I get the domain mismatch error:
**TypeError: Initial conditions cannot contain 'Variable' objects, but 'Variable(-0x1ec6c385b756daa4, Negative electrode potential [V], children=[], domains={'primary': ['negative electrode']})' found in initial conditions for 'Positive electrode potential [V]'**

The reason for this error : The boundary condition for the positive electrode is I_app/A/sigma_p. From the above 3 lines, I think the Battery voltage calcualtion (BV) is function of positive domain and negative domain, and since this is used in the boolean operation in "i_applied", I am getting the domain error. Please correct me if I am wrong?

While I understand this part, I am afraid I have no clue on how to fix this. Any suggestions?

**Reply by valentinsulzer** [reply]
There is some circularity in how you are defining variables. I don't think that current function you're trying is possible. When the voltage hits the check the current will drop which will mean the voltage will go back below the check and it might struggle to solve the model at that point.
You would have to either wait until experiments work with basic models, or define your own controller using `sim.step`, inputting a new current at each time step based on the value of the voltage.

**Reply by veera-visw** [reply]
@valentinsulzer  I see your point. I have not used sim.step before; which source code will be a good starting point to see how it is used? 

Thanks
